sam0 flashpage_write: correct translation from RIOT pages to CPU pages writing

fedepell opened the pull request:
### Contribution description
Solve erroneous writing on flash with sam0 boards because of wrong assumption of page sizing. Discovered on a saml21-xpro board.
A "RIOT page" is 256 bytes but a physical page on sam0 is 64, therefore the write has to be translated into four 64 writes.

### Testing procedure
The automated test `tests/periph_flashpage` can be used to verify the correctness of the PR.

### Issues/PRs references
Fixes: #10068

jcarrano commented on cpu/sam0_common/periph/flashpage.c:
```diff
@@ -103,7 +103,15 @@ void flashpage_write(int page, const void *data)
     /* write data to page */
     if (data != NULL) {
-        flashpage_write_raw(page_addr, data, FLASHPAGE_SIZE);
+      /* One RIOT page is 4 SAM0 flash pages (a row) as described in the file
```
The docs are nice, but I wonder if the number "4" can be derived from other values, and if you can provide some sort of check in case any value changes.

fedepell replied:
about removing that "4": I would give that a look in depth asap, but should we tackle it as a separate PR since it's something that was there already before (to make that change more traceable and not mixed into this one)?

jcarrano commented:
> I don't know if there is (or could be added?) an automated test for this (in theory it is simple, but is potentially destructive as writes directly a flash page).

Destructive in the sense that it wears flash? In that case flashing the test binary is just as destructive.
If you are worried about overwriting a part of the program, you can fix that by declaring a const array of page size, aligned to a page boundary and that should reserve a space that you can safely overwrite.

dylad commented:
@jcarrano 
What about` tests/periph_flashpage` ?
This test is intend to use this driver and play with the flash memory.

dylad commented:
>  I want to push for PRs that test themselves and contribute to enhance the RIOT test suite.

I understand your PoV but this is out of the scope of this PR.

aabadie commented:
> I want to push for PRs that test themselves and contribute to enhance the RIOT test suite. That means automated tests

I see this as a good candidate for its own PR => just add an `autotest` shell command that checks the different features of flashpage, call it from a Python script and parse the output.
But I also think this is out of the scope of this PR.

fedepell commented:
Hi,
Regarding a manual test of this PR the already present test tests/periph_flashpage that @dylad mentioned indeed shows openly the problem on the board:

-) Without patch:

```
ROM flash read write test
Please refer to the README.md for further information

Flash start addr:       0x00000000
Page size:              256
Number of pages:        1024
> 
> dump_local
Local page buffer:
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
> write 100
error: verification for page 100 failed
> read 100
Read flash page 100 into local page buffer
Local page buffer:
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
>
```

You see the write fails explicitly and just the last part of flash is written while the rest is 0xff which means just erased.

-) With patch applied:
```

> dump_local
Local page buffer:
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
> write 100
wrote local page buffer to flash page 100 at addr 0x6400
> read 100
Read flash page 100 into local page buffer
Local page buffer:
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
>
```

So seems to work (of course I just tested on my saml21-xpro).


Regarding automated test: I totally agree there should be some. I will try to study a bit myself the test part of RIOT (I'm rather new, need some time :) ) and see if I can contribute something in that direction!

Thanks for comments and insights!

aabadie commented:
@fedepell, have a look at the `shell` test application. There are shell commands defined (for example [start_test](https://github.com/RIOT-OS/RIOT/blob/master/tests/shell/main.c#L56)) and a [python script](https://github.com/RIOT-OS/RIOT/blob/master/tests/shell/tests/01-run.py) that call each of them and check their output using the pexpect module.
With pexpect, you can use regexp to finely parse the output.

The python script is run using the `test` target:

    make flash test

So the idea would to call the `test` shell command and parse its output from the python script.

fedepell commented:
@aabadie : thanks for the pointer! Please give a look at #10106 and let me know if it sounds good (or at least a good start :) ! Thanks!

aabadie commented:
There is an automated test now: #10106 is merged. Please rebase and adapt the testing procedure of this PR.

fedepell commented:
I tried to clean up a bit the PR this morning, hope it's getting better.

@jcarrano : I have removed the magic 4 and put a define instead and improved the comments (the old comment was actually stating the opposite). I also corrected a bit the code formatting.

About the fact that this "4" could be in principle any number: at the moment the code makes quite an assumption that this number is 4. To make this number easily changeable we should take into account that at least:
-) page erasing erases a row, so 4 at a time. In a generic case we should take this into account. If a multiple of 4 would be easy to implement (just loop over erasing pages), if not a multiple could be very complicated to implement (as partial pages cannot be erased, we would need to save parts of page)
-) data should be anyway contiguous. So for example 4 pages and 256 length, 8 and 512 and so on. If not then also higher level functions (such as flashpage_addr()) should be changed to take this into account (but we would be wasting flash).

I did a manual test on 8/512 (with two erases) and then it works but this is not a generic solution and I'm not sure it makes sense to implement one. 4 is quite hardware optimized (given that 4 pages on SAM0 are a row) and I don't see any performance improvement in going on higher multiples (given the size of flash we have, could make sense if we had very big flash memories).

Thanks!
F.

dylad commented:
I ran some tests on arduino-zero like board. automated tests work well.
@jcarrano @aabadie Do you have any more comments ?

dylad commented:
ping @jcarrano @aabadie

aabadie commented:
> Do you have any more comments ?

No comment on my side. @jcarrano, what's your opinion ?

fedepell commented:
anybody for an approving review? cannot wait to do remote flashing with this now that #10065 is going in ;)

dylad commented:
@fedepell Could you rebase this PR to the current master please ? I'll re-run a quick test then and ACK if it's all good :)

fedepell commented:
@dylad : done, rebased and checked locally on my saml. 

Thanks a lot for taking time to check for this! :)

dylad commented:
@fedepell I ran some tests with this PR and I have some issues : 
`test_last_raw` doesn't seem to work properly, I cannot see the pattern `"test12344321tset"` afterwards using `read 1023` (which is supposed to be the last flashpage of SAML21-XPRO)
Also, test_last is supposed to fill up the whole last flashpage. When I run `test_last` followed by `read 1023` I get the following output : 

```
2018-12-20 11:55:09,887 - INFO #  test_last
2018-12-20 11:55:09,894 - INFO # wrote local page buffer to last flash page
> read 1023
2018-12-20 11:55:17,430 - INFO #  read 1023
2018-12-20 11:55:17,434 - INFO # Read flash page 1023 into local page buffer
2018-12-20 11:55:17,436 - INFO # Local page buffer:
2018-12-20 11:55:17,444 - INFO # 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 
2018-12-20 11:55:17,453 - INFO #   a    b    c    d    e    f    g    h    i    j    k    l    m    n    o    p  
2018-12-20 11:55:17,461 - INFO # 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 
2018-12-20 11:55:17,469 - INFO #   q    r    s    t    u    v    w    x    y    z    a    b    c    d    e    f  
2018-12-20 11:55:17,478 - INFO # 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 
2018-12-20 11:55:17,486 - INFO #   g    h    i    j    k    l    m    n    o    p    q    r    s    t    u    v  
2018-12-20 11:55:17,494 - INFO # 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 
2018-12-20 11:55:17,502 - INFO #   w    x    y    z    a    b    c    d    e    f    g    h    i    j    k    l  
2018-12-20 11:55:17,511 - INFO # 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 
2018-12-20 11:55:17,519 - INFO #   m    n    o    p    q    r    s    t    u    v    w    x    y    z    a    b  
2018-12-20 11:55:17,527 - INFO # 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 
2018-12-20 11:55:17,535 - INFO #   c    d    e    f    g    h    i    j    k    l    m    n    o    p    q    r  
2018-12-20 11:55:17,544 - INFO # 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 
2018-12-20 11:55:17,552 - INFO #   s    t    u    v    w    x    y    z    a    b    c    d    e    f    g    h  
2018-12-20 11:55:17,560 - INFO # 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 
2018-12-20 11:55:17,569 - INFO #   i    j    k    l    m    n    o    p    q    r    s    t    u    v    w    x  
2018-12-20 11:55:17,577 - INFO # 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 
2018-12-20 11:55:17,585 - INFO #   y    z    a    b    c    d    e    f    g    h    i    j    k    l    m    n  
2018-12-20 11:55:17,594 - INFO # 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 
2018-12-20 11:55:17,602 - INFO #   o    p    q    r    s    t    u    v    w    x    y    z    a    b    c    d  
2018-12-20 11:55:17,610 - INFO # 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 
2018-12-20 11:55:17,618 - INFO #   e    f    g    h    i    j    k    l    m    n    o    p    q    r    s    t  
2018-12-20 11:55:17,627 - INFO # 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 
2018-12-20 11:55:17,635 - INFO #   u    v    w    x    y    z    a    b    c    d    e    f    g    h    i    j  
2018-12-20 11:55:17,643 - INFO # 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
2018-12-20 11:55:17,651 - INFO #   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
2018-12-20 11:55:17,659 - INFO # 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
2018-12-20 11:55:17,667 - INFO #   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
2018-12-20 11:55:17,675 - INFO # 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
2018-12-20 11:55:17,683 - INFO #   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
2018-12-20 11:55:17,691 - INFO # 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
2018-12-20 11:55:17,699 - INFO #   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
```

Finally, use the command `write 1023` which should copy the local page to the last page of the MCU gives me this : 

```
2018-12-20 12:00:02,353 - INFO #  write 1023
2018-12-20 12:00:02,356 - INFO # 0x1187
2018-12-20 12:00:02,358 - INFO # *** RIOT kernel panic:
2018-12-20 12:00:02,360 - INFO # FAILED ASSERTION.
2018-12-20 12:00:02,360 - INFO # 
2018-12-20 12:00:02,361 - INFO # *** halted.
2018-12-20 12:00:02,361 - INFO # 
```
Did I miss something ?

dylad commented:
I already tried but with no luck.

Here some output : 


```
> test_last_raw
2018-12-20 12:43:17,310 - INFO #  test_last_raw
2018-12-20 12:43:17,314 - INFO # wrote raw short buffer to last flash page
> read 1022
2018-12-20 12:43:21,139 - INFO #  read 1022
2018-12-20 12:43:21,143 - INFO # Read flash page 1022 into local page buffer
2018-12-20 12:43:21,145 - INFO # Local page buffer:
2018-12-20 12:43:21,153 - INFO # 0x60 0x60 0x63 0x64 0x21 0x22 0x23 0x20 0x20 0x22 0x22 0x20 0x64 0x62 0x65 0x70
2018-12-20 12:43:21,162 - INFO #   `    `    c    d    !    "    #              "    "         d    b    e    p
2018-12-20 12:43:21,170 - INFO # 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66
2018-12-20 12:43:21,178 - INFO #   q    r    s    t    u    v    w    x    y    z    a    b    c    d    e    f
2018-12-20 12:43:21,187 - INFO # 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76
2018-12-20 12:43:21,195 - INFO #   g    h    i    j    k    l    m    n    o    p    q    r    s    t    u    v
2018-12-20 12:43:21,203 - INFO # 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c
2018-12-20 12:43:21,211 - INFO #   w    x    y    z    a    b    c    d    e    f    g    h    i    j    k    l
2018-12-20 12:43:21,220 - INFO # 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62
2018-12-20 12:43:21,228 - INFO #   m    n    o    p    q    r    s    t    u    v    w    x    y    z    a    b
2018-12-20 12:43:21,236 - INFO # 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72
2018-12-20 12:43:21,244 - INFO #   c    d    e    f    g    h    i    j    k    l    m    n    o    p    q    r
2018-12-20 12:43:21,253 - INFO # 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68
2018-12-20 12:43:21,261 - INFO #   s    t    u    v    w    x    y    z    a    b    c    d    e    f    g    h
2018-12-20 12:43:21,269 - INFO # 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78
2018-12-20 12:43:21,278 - INFO #   i    j    k    l    m    n    o    p    q    r    s    t    u    v    w    x
2018-12-20 12:43:21,286 - INFO # 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e
2018-12-20 12:43:21,294 - INFO #   y    z    a    b    c    d    e    f    g    h    i    j    k    l    m    n
2018-12-20 12:43:21,303 - INFO # 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64
2018-12-20 12:43:21,311 - INFO #   o    p    q    r    s    t    u    v    w    x    y    z    a    b    c    d
2018-12-20 12:43:21,319 - INFO # 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74
2018-12-20 12:43:21,327 - INFO #   e    f    g    h    i    j    k    l    m    n    o    p    q    r    s    t
2018-12-20 12:43:21,336 - INFO # 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a
2018-12-20 12:43:21,344 - INFO #   u    v    w    x    y    z    a    b    c    d    e    f    g    h    i    j
2018-12-20 12:43:21,352 - INFO # 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a
2018-12-20 12:43:21,361 - INFO #   k    l    m    n    o    p    q    r    s    t    u    v    w    x    y    z
2018-12-20 12:43:21,369 - INFO # 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70
2018-12-20 12:43:21,377 - INFO #   a    b    c    d    e    f    g    h    i    j    k    l    m    n    o    p
2018-12-20 12:43:21,386 - INFO # 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66
2018-12-20 12:43:21,394 - INFO #   q    r    s    t    u    v    w    x    y    z    a    b    c    d    e    f
2018-12-20 12:43:21,402 - INFO # 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76
2018-12-20 12:43:21,410 - INFO #   g    h    i    j    k    l    m    n    o    p    q    r    s    t    u    v
> read 1023
2018-12-20 12:43:24,709 - INFO #  read 1023
2018-12-20 12:43:24,714 - INFO # Read flash page 1023 into local page buffer
2018-12-20 12:43:24,715 - INFO # Local page buffer:
2018-12-20 12:43:24,724 - INFO # 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70
2018-12-20 12:43:24,732 - INFO #   a    b    c    d    e    f    g    h    i    j    k    l    m    n    o    p
2018-12-20 12:43:24,740 - INFO # 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66
2018-12-20 12:43:24,749 - INFO #   q    r    s    t    u    v    w    x    y    z    a    b    c    d    e    f
2018-12-20 12:43:24,757 - INFO # 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76
2018-12-20 12:43:24,765 - INFO #   g    h    i    j    k    l    m    n    o    p    q    r    s    t    u    v
2018-12-20 12:43:24,774 - INFO # 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c
2018-12-20 12:43:24,782 - INFO #   w    x    y    z    a    b    c    d    e    f    g    h    i    j    k    l
2018-12-20 12:43:24,790 - INFO # 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62
2018-12-20 12:43:24,798 - INFO #   m    n    o    p    q    r    s    t    u    v    w    x    y    z    a    b
2018-12-20 12:43:24,807 - INFO # 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72
2018-12-20 12:43:24,815 - INFO #   c    d    e    f    g    h    i    j    k    l    m    n    o    p    q    r
2018-12-20 12:43:24,823 - INFO # 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68
2018-12-20 12:43:24,832 - INFO #   s    t    u    v    w    x    y    z    a    b    c    d    e    f    g    h
2018-12-20 12:43:24,840 - INFO # 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78
2018-12-20 12:43:24,848 - INFO #   i    j    k    l    m    n    o    p    q    r    s    t    u    v    w    x
2018-12-20 12:43:24,856 - INFO # 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e
2018-12-20 12:43:24,865 - INFO #   y    z    a    b    c    d    e    f    g    h    i    j    k    l    m    n
2018-12-20 12:43:24,873 - INFO # 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64
2018-12-20 12:43:24,881 - INFO #   o    p    q    r    s    t    u    v    w    x    y    z    a    b    c    d
2018-12-20 12:43:24,890 - INFO # 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74
2018-12-20 12:43:24,898 - INFO #   e    f    g    h    i    j    k    l    m    n    o    p    q    r    s    t
2018-12-20 12:43:24,906 - INFO # 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a
2018-12-20 12:43:24,914 - INFO #   u    v    w    x    y    z    a    b    c    d    e    f    g    h    i    j
2018-12-20 12:43:24,923 - INFO # 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff
2018-12-20 12:43:24,930 - INFO #   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??
2018-12-20 12:43:24,939 - INFO # 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff
2018-12-20 12:43:24,946 - INFO #   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??
2018-12-20 12:43:24,955 - INFO # 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff
2018-12-20 12:43:24,962 - INFO #   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??
2018-12-20 12:43:24,971 - INFO # 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff
2018-12-20 12:43:24,978 - INFO #   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??
```

fedepell commented:
@dylad : I will try to check tonight or tomorrow with my HW by my side, sorry for this!

fedepell commented:
@dylad : sorry for the delay, didn't have the board with me in daytime today :)

So all clear now! The _test_last_raw_ does the write with _flashpage_write_raw_ but does not erase the page first, therefore the write de facto fails (but there is no reporting back about this, since you can only know it failed if you re-read and compare). If you first erase the page then everything works fine.

This explains why _test_last_ works: in that case _flashpage_write_ is called which first erases the page, then writes with _flashpage_write_raw_ (in the new loop that was modified by this PR).



So the code of this PR should be fine I believe, what should be fixed is the test (put an erase before the call to _flash_write_raw_). Question: should I do this in this PR (with a separate commit) or a separate one or what do you suggest is the best?


Here is the output showing what I wrote above:

```
> main(): This is RIOT! (Version: 2018.10-RC1-622-ga0054-sam0_flashwrite)
ROM flash read write test

Please refer to the README.md for main(): This is RIOT! (Version: 2018.10-RC1-622-ga0054-sam0_flashwrite)
ROM flash read write test

Please refer to the README.md for further information

Flash start addr:       0x00000000
Page size:              256
Number of pages:        1024
> test_last
wrote local page buffer to last flash page
> read 1022
Read flash page 1022 into local page buffer
Local page buffer:
0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 
  a    b    c    d    e    f    g    h    i    j    k    l    m    n    o    p  
0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 
  q    r    s    t    u    v    w    x    y    z    a    b    c    d    e    f  
0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 
  g    h    i    j    k    l    m    n    o    p    q    r    s    t    u    v  
0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 
  w    x    y    z    a    b    c    d    e    f    g    h    i    j    k    l  
0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 
  m    n    o    p    q    r    s    t    u    v    w    x    y    z    a    b  
0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 
  c    d    e    f    g    h    i    j    k    l    m    n    o    p    q    r  
0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 
  s    t    u    v    w    x    y    z    a    b    c    d    e    f    g    h  
0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 
  i    j    k    l    m    n    o    p    q    r    s    t    u    v    w    x  
0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 
  y    z    a    b    c    d    e    f    g    h    i    j    k    l    m    n  
0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 
  o    p    q    r    s    t    u    v    w    x    y    z    a    b    c    d  
0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 
  e    f    g    h    i    j    k    l    m    n    o    p    q    r    s    t  
0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 
  u    v    w    x    y    z    a    b    c    d    e    f    g    h    i    j  
0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 
  k    l    m    n    o    p    q    r    s    t    u    v    w    x    y    z  
0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 
  a    b    c    d    e    f    g    h    i    j    k    l    m    n    o    p  
0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 
  q    r    s    t    u    v    w    x    y    z    a    b    c    d    e    f  
0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 
  g    h    i    j    k    l    m    n    o    p    q    r    s    t    u    v  
> test_last_raw
wrote raw short buffer to last flash page
> read 1022
Read flash page 1022 into local page buffer
Local page buffer:
0x60 0x60 0x63 0x64 0x21 0x22 0x23 0x20 0x20 0x22 0x22 0x20 0x64 0x62 0x65 0x70 
  `    `    c    d    !    "    #              "    "         d    b    e    p  
0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 
  q    r    s    t    u    v    w    x    y    z    a    b    c    d    e    f  
0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 
  g    h    i    j    k    l    m    n    o    p    q    r    s    t    u    v  
0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 
  w    x    y    z    a    b    c    d    e    f    g    h    i    j    k    l  
0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 
  m    n    o    p    q    r    s    t    u    v    w    x    y    z    a    b  
0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 
  c    d    e    f    g    h    i    j    k    l    m    n    o    p    q    r  
0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 
  s    t    u    v    w    x    y    z    a    b    c    d    e    f    g    h  
0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 
  i    j    k    l    m    n    o    p    q    r    s    t    u    v    w    x  
0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 
  y    z    a    b    c    d    e    f    g    h    i    j    k    l    m    n  
0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 
  o    p    q    r    s    t    u    v    w    x    y    z    a    b    c    d  
0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 
  e    f    g    h    i    j    k    l    m    n    o    p    q    r    s    t  
0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 
  u    v    w    x    y    z    a    b    c    d    e    f    g    h    i    j  
0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 
  k    l    m    n    o    p    q    r    s    t    u    v    w    x    y    z  
0x61 0x62 0x63 0x64 0x65 0x66 0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 
  a    b    c    d    e    f    g    h    i    j    k    l    m    n    o    p  
0x71 0x72 0x73 0x74 0x75 0x76 0x77 0x78 0x79 0x7a 0x61 0x62 0x63 0x64 0x65 0x66 
  q    r    s    t    u    v    w    x    y    z    a    b    c    d    e    f  
0x67 0x68 0x69 0x6a 0x6b 0x6c 0x6d 0x6e 0x6f 0x70 0x71 0x72 0x73 0x74 0x75 0x76 
  g    h    i    j    k    l    m    n    o    p    q    r    s    t    u    v  
> erase 1022
successfully erased page 1022 (addr 0x3fe00)
> test_last_raw
wrote raw short buffer to last flash page
> read 1022
Read flash page 1022 into local page buffer
Local page buffer:
0x74 0x65 0x73 0x74 0x31 0x32 0x33 0x34 0x34 0x33 0x32 0x31 0x74 0x73 0x65 0x74 
  t    e    s    t    1    2    3    4    4    3    2    1    t    s    e    t  
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 0xff 
  ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ??   ?? 
```

Thanks!

dylad commented:
@fedepell I'm fine with another commit here since this is not a 'real' bug. Thanks for re-testing. I'll retry tomorrow on hardware.

fedepell commented:
@dylad : added commit with the modification to the test_last_raw command. Hope everything will be fine now! :)

dylad commented:
@fedepell your fix works like a charm ! However I still get a kernel panic if I try to use the` write 1023 ` command. I think it would be better to fix it now otherwise you will get weird behaviour later while playing with bootloader.

dylad commented:
https://github.com/RIOT-OS/RIOT/blob/719612092b02bdc2976604e1516cf483bd26e868/cpu/sam0_common/periph/flashpage.c#L70-L72
I think the `< `condition should be `<=` otherwise we cannot write the very last byte of flash.

fedepell commented:
@dylad : thanks for testing!

You are absolutely correct, the assert needs to be >= to be able to write until exactly the last byte! I corrected, tested it succesfully and added a commit.

Let me know if we should squash or not (would in principle say to keep the 3 commits separated as they deal with different issues, so it is easier to track in the future, but let me know!)

dylad commented:
@fedepell Keep it that way, it's perfect. I'll re-test again.

fedepell commented:
(made the comment shorter for Murdock ;) )

dylad commented on cpu/sam0_common/periph/flashpage.c:
```diff
@@ -103,7 +103,16 @@ void flashpage_write(int page, const void *data)
+         * The erasing is done once as a full row is always reased.
+         */
+        for (unsigned curpage = 0; curpage < FLASHPAGE_PAGES_PER_ROW; curpage++) {
+            flashpage_write_raw(page_addr + (curpage * NVMCTRL_PAGE_SIZE / 4), (void *) ((uint32_t *) data + (curpage * NVMCTRL_PAGE_SIZE / 4)), NVMCTRL_PAGE_SIZE);
```
Please stay within the 80 characters limit per lines. @aabadie will murder me if we merge this ;)

fedepell replied:
Just one question: is 80 very strict? To keep it nicely readadble I would keep it as:

```
        for (unsigned curpage = 0; curpage < FLASHPAGE_PAGES_PER_ROW; curpage++) {
            flashpage_write_raw(page_addr + (curpage * NVMCTRL_PAGE_SIZE / 4),
                                (void *) ((uint32_t *) data + (curpage * NVMCTRL_PAGE_SIZE / 4)),
                                NVMCTRL_PAGE_SIZE);
        }
```
But the line with the second parameter is 98 sadly. Otherwise I can split that too to make it <= 80 but will become somehow less readable, something like:
```
        for (unsigned curpage = 0; curpage < FLASHPAGE_PAGES_PER_ROW; curpage++) {
            flashpage_write_raw(page_addr + (curpage * NVMCTRL_PAGE_SIZE / 4),
                                (void *) ((uint32_t *) data + 
                                (curpage * NVMCTRL_PAGE_SIZE / 4)),
                                NVMCTRL_PAGE_SIZE);
        }
```

aabadie replied:
Having it less than 100 is ok. The most important is to keep the code readable.

fedepell replied:
Thanks for the indication! I committed now the first version which is much more readable and still better than the original long line. Hope it's fine now!

dylad commented on cpu/sam0_common/include/cpu_conf.h:
```diff
@@ -43,8 +43,10 @@ extern "C" {
 #define FLASHPAGE_SIZE             (256U)
-/* one SAM0 row contains 4 SAM0 pages -> 4x the amount of RIOT flashpages */
-#define FLASHPAGE_NUMOF            (FLASH_NB_OF_PAGES / 4)
+/* one SAM0 row contains 4 SAM0 pages, so 4 SAM0 pages contain the amount of a RIOT flashpage */
```
same here

fedepell replied:
ok no problem for this!

dylad commented:
Could you squash the last commit so we keep 3 commits please ?

fedepell commented:
> Could you squash the last commit so we keep 3 commits please ?

Done!

dylad commented:
@fedepell Thanks for your good work !
